crates/autopilot/src/domain/settlement/mod.rs

```diff
@@ -2,39 +2,151 @@
+                    .map_err(|err| Error::OrderUidRecover(err, auction_id))?,
+                eth::Asset {
+                    token: sell_token.into(),
+                    amount: trade.3.into(),
+                },
+                eth::Asset {
+                    token: buy_token.into(),
+                    amount: trade.4.into(),
+                },
+                flags.side(),
+                trade.9.into(),
+                trade::Prices {
+                    uniform: trade::ClearingPrices {
+                        sell: clearing_prices[uniform_sell_token_index],
+                        buy: clearing_prices[uniform_buy_token_index],
+                    },
+                    custom: trade::ClearingPrices {
+                        sell: clearing_prices[sell_token_index],
+                        buy: clearing_prices[buy_token_index],
+                    },
+                },
+            ));
+        }
+
+        let auction = persistence.get_auction(auction_id).await?;
+        let solution = persistence.get_competition_solution(auction_id).await?;
+
+        let score = domain::competition::Score::new(
+            trades
+                .iter()
+                .map(|trade| {
+                    trade.score(
+                        &auction.prices,
+                        auction
+                            .fee_policies
+                            .get(trade.order_uid())
+                            .map(|value| value.as_slice())
+                            .unwrap_or_default(),
+                    )
+                })
+                .sum::<Result<eth::Ether, trade::Error>>()
+                .map_err(error::Score::from)?,
+        )
+        .map_err(error::Score::from)?;
+
+        if score != solution.score() {
+            return Err(Error::Score(error::Score::Missmatch));
+        }
+
+        // TODO implement the rest of the checks
+
+        Ok(Self { trades, auction })
     }
 }
 
-#[derive(Debug, thiserror::Error)]
-pub enum Error {
-    #[error(transparent)]
-    Blockchain(#[from] infra::blockchain::Error),
-    #[error(transparent)]
-    Solution(#[from] solution::Error),
+mod error {
+    use super::*;
+
+    #[derive(Debug, thiserror::Error)]
+    pub enum Error {
+        #[error(transparent)]
+        Decoding(#[from] tokenized::error::Decoding),
+        #[error("failed to recover order uid {0} for auction {1}")]
+        OrderUidRecover(tokenized::error::Uid, crate::domain::auction::Id),
+        #[error(transparent)]
+        Score(#[from] Score),
+        #[error(transparent)]
+        Auction(#[from] infra::persistence::error::Auction),
+        #[error(transparent)]
+        Solution(#[from] infra::persistence::error::Solution),
+    }
+    #[derive(Debug, thiserror::Error)]
+    pub enum Score {
+        /// Per CIP38, zero score solutions are rejected.
+        #[error(transparent)]
+        Zero(#[from] domain::competition::ZeroScore),
```

Comment:
Is there any value in wrapping this empty struct here? Why not just have it `Score::Zero` to be less of a nested error mess?

I see you are frequently creating new error structs that simply wrap the inner error types with an automatic conversion a la `#[from]`. While this is very convenient and avoids us having to think about the errors too much, I think it will cause confusion and a mess down the line. I think we should think as carefully about the error types (what errors make sense at this level and would be handled differently) instead of aggregating all potential suberror types in an somewhat automatically generated super error type.